editoast/src/main.rs

```diff
@@ -221,17 +205,36 @@ fn clear(args: ClearArgs, pg_config: PostgresConfig) -> Result<(), Box<dyn Error
     Ok(())
 }
 
+fn cache<'a>(
+    infra_id: i32,
+    conn: &'a PgConnection,
+    infra_caches: &'a State<Arc<CHashMap<i32, InfraCache>>>,
+) -> ReadGuard<'a, i32, InfraCache> {
+    match infra_caches.get(&infra_id) {
+        Some(infra_cache) => infra_cache,
+        None => {
+            let infra_cache = InfraCache::load(&conn, infra_id);
+            infra_caches.insert_new(infra_id, infra_cache);
+            infra_caches.as_ref().get(&infra_id).unwrap()
+        }
+    }
+}
```

Comment:
- This function should be moved to a more proper place.
- You should implement a variant that returns a `WriteGuard` (to be able to get a mutable version of the infra_cache)

--------

Here is a more clean way to implement it.

```rust
/// This function tries to get the infra from the cache, if it fails, it loads it from the database
fn get_or_load_infra_cache<'a>(
    conn: &PgConnection,
    infra_caches: &'a CHashMap<i32, InfraCache>,
    infra_id: i32,
) -> ReadGuard<'a, i32, InfraCache> {
    // Cache hit
    if let Some(infra_cache) = infra_caches.get(&infra_id) {
        return infra_cache;
    }
    // Cache miss
    infra_caches.insert_new(infra_id, InfraCache::load(conn, infra_id));
    infra_caches.get(&infra_id).unwrap()
}
```